Adding Python Django test docs (#7516)
microsoft/vscode-docs: docs/python/testing.md

![Discovery failure error messaged displayed in the Test Explorer](images/testing/test-discovery-error.png)

### Django unittests

The Python extension also offers support for running Django unittests! However, if you are writing tests for your Django application, there are a few additional steps needed in order to have your tests discovered.

First, ensure you have correctly set `"python.testing.unittestEnabled": true,` in your `settings.json` file, then follow the steps below to add `MANAGE_PY_PATH` as an environment variable:

- Create a file named `.env` at the root of your project.
- Add `MANAGE_PY_PATH='path-string-to-manange.py-path'` to the `.env` file, replacing `path-string-to-manange.py-path` with the path to your application's `manage.py` file (Note: you can copy the path by right clicking on the file and selecting **Copy Path**).
- Add the following to your `settings.json` file `"python.envFile": "${workspaceFolder}/.env”` to incorporate this file in your environment (edit path as necessary if `.env` is not at root).
- Add Django args to `"python.testing.unittestArgs": []` in `settings.json` as needed, and remove any arguments that are not compatible with Django.

> Navigate to the Testing Panel, and refresh the tree-view to see your Django tests populate!

Author: ```suggestion

Navigate to the Testing Panel, and refresh the tree-view to see your Django tests populate!
```

![Discovery failure error messaged displayed in the Test Explorer](images/testing/test-discovery-error.png)

### Django unittests

> The Python extension also offers support for running Django unittests! However, if you are writing tests for your Django application, there are a few additional steps needed in order to have your tests discovered.

Reviewer: ```suggestion
The Python extension also offers support for running Django unittests! You get your Django tests discovered with only a few additional set up steps.
```

> First, ensure you have correctly set `"python.testing.unittestEnabled": true,` in your `settings.json` file. Follow the steps below to add `MANAGE_PY_PATH` as an environment variable:

Reviewer 2: then follow the steps maybe

> First, ensure you have correctly set `"python.testing.unittestEnabled": true,` in your `settings.json` file. Follow the steps below to add `MANAGE_PY_PATH` as an environment variable:

Author: ```suggestion
First, ensure you have correctly set `"python.testing.unittestEnabled": true,` in your `settings.json` file, then follow the steps below to add `MANAGE_PY_PATH` as an environment variable:
```

- Create a file named `.env` at the root of your project.

> - Add `MANAGE_PY_PATH='path-string-to-manange.py-path'` to the `.env` file, replacing `path-string-to-manange.py-path` with the path to your application's `manage.py` file (Note: you can copy the path by right clicking on the file in the File Explorer and selecting **Copy Path**).

Reviewer 3: ```suggestion
- Add `MANAGE_PY_PATH='path-string-to-manage.py-path'` to the `.env` file, replacing `path-string-to-manage.py-path` with the path to your application's `manage.py` file (Note: you can copy the path by right clicking on the file in the File Explorer and selecting **Copy Path**).
```

> - Add `MANAGE_PY_PATH='path-string-to-manange.py-path'` to the `.env` file, replacing `path-string-to-manange.py-path` with the path to your application's `manage.py` file (Note: you can copy the path by right clicking on the file and selecting **Copy Path**).

Reviewer: ```suggestion
- Add `MANAGE_PY_PATH='path-string-to-manange.py-path'` to the `.env` file, replacing `path-string-to-manange.py-path` with the path to your application's `manage.py` file (Note: you can copy the path by right clicking on the file in the File Explorer and selecting **Copy Path**).
```

- Add the following to your `settings.json` file `"python.envFile": "${workspaceFolder}/.env”` to incorporate this file in your environment (edit path as necessary if `.env` is not at root).

> 

Reviewer 2: some users have run into the problem that we populate `python.testing.unittestArgs` if they go through the UI setup flow right now and these args don't work for Django. Could we add somewhere to edit that variable to remove args and add any Django necessary args

Author: Moved this reminder up to the steps

> 

Author: ```suggestion
- Add Django args to `"python.testing.unittestArgs": []` in `settings.json` as needed, and remove any arguments that are not compatible with Django.
```

> - Add the following to your `settings.json` file `"python.envFile": "${workspaceFolder}/.env”` to incorporate this file in your environment (edit path as necessary if `.env` is not at root).

Reviewer: ```suggestion
- Add  `"python.envFile": "${workspaceFolder}/.env”` to your `settings.json` file, so the Python extension can load the environment variables in this file when running and discovering tests (edit path as necessary if `.env` is not at root).
```

> - Add  `"python.envFile": "${workspaceFolder}/.env”` to your `settings.json` file, so the Python extension can load the environment variables in this file when running and discovering tests (edit path as necessary if `.env` is not at root).

Reviewer: ```suggestion
- Add  `"python.envFile": "${workspaceFolder}/.env"` to your `settings.json` file, so the Python extension can load the environment variables in this file when running and discovering tests (edit path as necessary if `.env` is not at root).
```

Navigate to the Testing Panel, and refresh the tree-view to see your Django tests populate!

In the case your tests are not discovered, try the following troubleshooting steps:

> - Add `"python.testing.unittestArgs": []` in `settings.json`

Reviewer 2: yeah like this but it is so common it feels almost required?

> - Add `"python.testing.unittestArgs": []` in `settings.json`

Author: ```suggestion
```

> Navigate to the Testing Panel, and refresh the tree-view to see your Django tests populate!

Reviewer: ```suggestion
Navigate to the Testing Panel, and refresh the tree-view to have your Django tests populated!
```

Reviewer: or "displayed" instead of populated. Main suggestion here is to replace "see"

> Navigate to the Testing Panel, and refresh the tree-view to have your Django tests displayed!

Reviewer 4: ```suggestion
Navigate to the Testing view, and select the **Refresh Tests** button to have your Django tests displayed!
```

> - Look at the Python Output Panel for any error messages that may have triggered.

Reviewer 2: do we want to remind people how to change logging verbosity or is that elsewhere in our docs?

> In the case your tests are not discovered, try the following troubleshooting steps:

Reviewer: I'm wondering if it'd be worth adding a Troubleshooting section like we do for [Linting ](https://code.visualstudio.com/docs/python/linting#_troubleshooting-linting) and [Formatting](https://code.visualstudio.com/docs/python/formatting#_troubleshoot-formatting)?

Author: Hmmm I am not opposed to it, however I am not sure if it would fit in the same table/labels as they have there since the issue would all be the tests aren't getting discovered and there are just a list of things to try. Are you thinking just having one row where the issue is "Tests are not getting discovered" and then list the possible solutions together?

Author: We could try framing it like this (with some cleaning up lol). 
| Problem | solution |
|----|----|
| There are configuration errors | Look at the Python Output channel for any error messages that may have triggered | 
| Commands that are being run are not the ones intended | Try running commands in the Command Line, then transfer the same command to how it is represented in VS Code. For example, you may run `python manage.py test --arg` in the Command Line which is represented by setting `MANAGE_PY_PATH='./manage.py'` and `"python.testing.unittestArgs": [--arg]` in VS Code. Check the output as it should print the commands that are actually ran. |
| The path to manage.py may be incorrect. | Attempt using the absolute path for `manage.py`, if you initially tried using the relative path. |

Reviewer 2: Yeah I think it is hard to say if these problems will appear differently for users- I think they will just see nothing show up or and error and multiple steps could be the solution for that

> - Look at the Python Output Panel for any error messages that may have triggered.

Reviewer: ```suggestion
- Search for error messages in the **Python** output panel.
```

> - Try running commands in the Command Line, then transfer the same command to how it is represented in VS Code. For example, you may run `python manage.py test --arg` in the Command Line which is represented by setting `MANAGE_PY_PATH='./manage.py'` and `"python.testing.unittestArgs": [--arg]` in VS Code. Check the output as it should print the commands that are actually ran.

Reviewer 4: ```suggestion
- Try running commands in the command line, then transfer the same command to how it is represented in VS Code. For example, you may run `python manage.py test --arg` in the command line, which is represented by setting `MANAGE_PY_PATH='./manage.py'` and `"python.testing.unittestArgs": [--arg]` in VS Code. Check the output as it should print the commands that are actually ran.
```

Reviewer: That's a good point. The target users of this feature are more "power" users of Django, so they would likely know the commands to run in the terminal (which could be the one in VS Code or an external one, shouldn't matter). Translating it into VS Code would ideally be following the set up steps and leveraging the Python extension's testing args setting if needed. 

Based on this comment I made a couple of changes:
- Changed command line -> terminal
- Moved the whole Django subsection into its own section, similar to what we've done to highlight IntelliSense for pytest
- Added a troubleshooting subsection to the Django unittest section, adding a few more details and linking to the Django docs on how to run tests in the terminal

- Attempt using the absolute path for `manage.py`, if you initially tried using the relative path.

> 

Reviewer 2: Thoughts on adding a known limitations section? Like right now custom django test runners do not work since it conflicts with our design as of now. We can also just direct them to the github as there will be issues there or know that is where they will go- idk depends on how we usually surface those types of things

Author: If would say if we don't ever plan on addressing those limitations then it is ok to document them, but if we might address them in the future we can just leave it be. I would say people typically would go to repo to make a feature request so it will be documented there.

Reviewer 2: yeah we hope to fix it so that sounds good to just leave it out

1. Set `"python.testing.unittestEnabled": true,` in your `settings.json`.
2. Add `MANAGE_PY_PATH` as an environment variable:
   1. Create a file named `.env` at the root of your project.
   2. Add `MANAGE_PY_PATH='path-string-to-manage.py-path'` to the `.env` file, replacing `path-string-to-manage.py-path` with the path to your application's `manage.py` file.
        > **Tip**: you can copy the path by right clicking on the file in the Explorer view and selecting **Copy Path**.
    3. Add  `"python.envFile": "${workspaceFolder}/.env"` to your `settings.json` file, so the Python extension can load the environment variables in this file when running and discovering tests (edit path as necessary if `.env` is not at root).
    4. Add Django test arguments to `"python.testing.unittestArgs": []` in `settings.json` as needed, and remove any arguments that are not compatible with Django.

Navigate to the Testing view, and select the **Refresh Tests** button to have your Django tests displayed!

### Troubleshooting
If your Django unit tests are not showing in the Testing view, try the following troubleshooting steps:

> - Search for error messages in the **Python** output panel. They may provide helpful hints as to why your tests are not being discovered.

Reviewer 4: ```suggestion
- Search for error messages in the **Python** Output panel. They might provide a hint as to why your tests are not being discovered.
```

> - Try [running the Django tests in the terminal](https://docs.djangoproject.com/en/dev/topics/testing/overview/#running-tests). Then "translate" the same command into VS Code settings. For example, if you run `python manage.py test --arg` in the terminal, you would add `MANAGE_PY_PATH='./manage.py'` to a `.env` file, and set `"python.testing.unittestArgs": [--arg]` in the VS Code's settings. You can also check the **Python** output panel as it displays the commands that are ran by the Python extension.

Reviewer 4: ```suggestion
- Try to [run the Django tests in the terminal](https://docs.djangoproject.com/en/dev/topics/testing/overview/#running-tests). Then "translate" the same command into VS Code settings.
    For example, if you run `python manage.py test --arg` in the terminal, you would add `MANAGE_PY_PATH='./manage.py'` to a `.env` file, and set `"python.testing.unittestArgs": [--arg]` in the VS Code settings. 
    
    Alternatively, you can also find the commands that are run by the Python extension in the **Python** Output panel.
```

> - Attempt using the absolute path for `manage.py` when setting the `MANAGE_PY_PATH` environment variable, if you initially tried using the relative path.

Reviewer 4: ```suggestion
- Use the absolute path to the `manage.py` file when setting the `MANAGE_PY_PATH` environment variable, if you initially used the relative path.
```

|  **Testing: Focus on Test Explorer View** | Open the Test Explorer view. Similar to **Testing: Focus on Python View** on versions prior to 2021.9.
|  **Test: Stop Refreshing Tests** | Cancel test discovery. |

## Django unit tests

The Python extension also offers support for discovering and running Django unit tests! You can get your Django tests discovered with only a few additional setup steps:

1. Set `"python.testing.unittestEnabled": true,` in your `settings.json` [file](/docs/getstarted/settings.md#settingsjson).
2. Add `MANAGE_PY_PATH` as an environment variable:
    1. Create a `.env` file at the root of your project.
    2. Add `MANAGE_PY_PATH='<path-to-manage.py>'` to the `.env` file, replacing `<path-to-manage.py>` with the path to your application's `manage.py` file.
        > **Tip**: you can copy the path by right clicking on the file in the Explorer view and selecting **Copy Path**.

>     4. Add Django test arguments to `"python.testing.unittestArgs": []` in the `settings.json` [file](/docs/getstarted/settings.md#settingsjson) as needed, and remove any arguments that are not compatible with Django.

Author: ```suggestion
    3. Add Django test arguments to `"python.testing.unittestArgs": []` in the `settings.json` [file](/docs/getstarted/settings.md#settingsjson) as needed, and remove any arguments that are not compatible with Django.
```

Author: Should this step be nested or represented top-level #3

Reviewer 4: It should probably be top-level #3 because it is not part of setting that environment variable.

> > **Note**: By default, the Python extension looks for and loads `.env` files at the project root. If your `.env` file is not at the project root or you are using [VS Code variable substitution](https://code.visualstudio.com/docs/editor/variables-reference), add `"python.envFile": "${workspaceFolder}/<path-to-.env>"` to your `settings.json` [file](/docs/getstarted/settings.md#settingsjson), so the Python extension can load the environment variables in this file when running and discovering tests. See our [Python environment variables](https://code.visualstudio.com/docs/python/environments#_environment-variables) docs for more information on environment variables.

Author: @luabud @ntrogh I moved this step to a note since this step is not needed to get it working by default. Please suggest edits as needed!

cc/ @eleanorjboyd

Reviewer 4: @cwebster-99 Merging this PR for today's release. I'll make a follow-on PR for some minor edits after the release.

> > **Note**: By default, the Python extension looks for and loads `.env` files at the project root. If your `.env` file is not at the project root or you are using [VS Code variable substitution](https://code.visualstudio.com/docs/editor/variables-reference), add `"python.envFile": "${workspaceFolder}/<path-to-.env>"` to your `settings.json` [file](/docs/getstarted/settings.md#settingsjson), so the Python extension can load the environment variables in this file when running and discovering tests. See our [Python environment variables](https://code.visualstudio.com/docs/python/environments#_environment-variables) docs for more information on environment variables.

Reviewer 4: ```suggestion
> **Note**: By default, the Python extension looks for and loads `.env` files at the project root. If your `.env` file is not at the project root or you are using [VS Code variable substitution](/docs/editor/variables-reference.md), add `"python.envFile": "${workspaceFolder}/<path-to-.env>"` to your `settings.json` [file](/docs/getstarted/settings.md#settingsjson). This enables the Python extension to load the environment variables from this file when running and discovering tests. Get more info about [Python environment variables](/docs/python/environments.md#_environment-variables).
```
